src/ci/run.sh

```diff
@@ -117,7 +117,10 @@ else
 
   # In general we always want to run tests with LLVM assertions enabled, but not
   # all platforms currently support that, so we have an option to disable.
-  if [ "$NO_LLVM_ASSERTIONS" = "" ]; then
+  # FIXME: the llvm in `rustc-dev` doesn't match the llvm in `rust-dev` if assertions are enabled
```

Comment:
Say more? Are we not downloading the alt rustc-dev component?

Poking at the code looks like no - should be easy to fix though?

As-is does this PR remove coverage for llvm assertions on x86_64-gnu? Or are those covered by some other runner?

Reply:
good catch, thank you!! you were right, this was simple to fix :) done in https://github.com/rust-lang/rust/pull/112143/commits/3afc4dc297f187021bc76c86d92928e07046eaa2 (and i think this should avoid regressing assertions after https://github.com/rust-lang/rust/pull/112143/commits/44857ad7c9c79e6adf284615b96d94244b059e6b).

Reply:
hmm, i am no longer sure this is a good idea. the alt builders also set parallel-compiler: https://github.com/rust-lang/rust/pull/112143#issuecomment-1627473887
and we can't use the llvm from `rust-dev` unconditionally because that breaks rustdoc: https://github.com/rust-lang/rust/pull/112143#issuecomment-1575100950
so if we turn this on here, we would be testing parallel-compiler for the first time in download-rustc CI (as seen from the list of linked PRs here, it's never been tested before and had a bunch of failures). that makes me very nervous. at a minimum we should probably run it by the parallel-compiler WG, but even if they think it's ok, i don't think testing it *only* when download-rustc is enabled is a good idea - at least we should have a builder that tests it normally too. 

so my preference for now would be to disable LLVM assertions for the x86_64-gnu builder. we still test assertions on other builders like `i686-gnu`.

Reply:
Hm, why do we enable parallel-compiler on dist-alt? The LLVM assertions + parallel-compiler combination seems to be of dubious value. Is it enabled there just to check that it builds? If so, can we move it somewhere else (preferably non-dist)?

Reply:
It's enabled there because we wanted somewhat broad coverage and ability for testing random commits, IIRC, e.g. for bisection and such. We already check that it builds (at least x.py check) in the mingw-check builder that runs on PRs and in CI, I believe. I don't think it's hard needed though, if it's giving us some kind of problem we can remove the parallel compilation. I don't understand what that problem is though (i.e., how that contaminates the libLLVM produced there).

> so my preference for now would be to disable LLVM assertions for the x86_64-gnu builder. we still test assertions on other builders like i686-gnu.

I am confused by this - seems like the right step is to disable parallel compilation or some similar thing? Not remove assertion coverage from alt builds, where ~99% of the "point" is those assertions.

>  we can't use the llvm from rust-dev unconditionally because that breaks rustdoc

I'm sure I don't yet understand the linking problems with rustdoc, but it seems like they ought to be solvable? I'm confused on what is causing this to happen only in CI, since I know lots of folks are using download-rustc (or so I assume).

Reply:
>  I'm confused on what is causing this to happen only in CI

at the time i wrote the comment, this PR was the only place that tested the combination of download-rustc and llvm assertions.

> seems like the right step is to disable parallel compilation or some similar thing?

that would work, but requires it to be disabled in the -alt builder. whether we use the alt artifacts for rustc has to match whether we use them for LLVM: https://github.com/rust-lang/rust/pull/112143#issuecomment-1575100950